msprime/intervals.py

```diff
@@ -229,26 +342,104 @@ def uniform_map(cls, length, rate, num_loci=None):
         return cls([0, length], [rate, 0], num_loci=num_loci)
 
     @classmethod
-    def read_hapmap(cls, filename):
+    def read_hapmap(cls, filename, sequence_length=None):
         """
-        Parses the specified file in HapMap format.
+        Parses the specified file in HapMap format. These files must be
+        white-space-delimited, and contain a single header line (which is
+        ignored), and then each subsequent line contains the starting position
+        and recombination rate for the segment from that position (inclusive)
+        to the starting position on the next line (exclusive). Starting
+        positions of each segment are given in units of bases, and
+        recombination rates in centimorgans/Megabase. The first column in this
+        file is checked , as are additional columns after the third (Position is
+        assumed to be the second column, and Rate is assumed to be the third).
+        If the first starting position is not equal to zero, then a
+        zero-recombination region is inserted at the start of the chromosome.
 
         .. warning::
             This method is deprecated, use the module-level
-            :func:`read_hapmap` function instead.
+            :func:`read_hapmap` function instead, which uses the genetic position
+            in centiMorgans rather than the rate column.
 
         :param str filename: The name of the file to be parsed. This may be
-            in plain text or gzipped plain text.
+            in plain text or gzipped plain text. It may also be a file object.
+        :param float sequence_length: The .
         :return: A RecombinationMap object.
         """
         warnings.warn(
             "RecombinationMap.read_hapmap() is deprecated. "
             "Use msprime.read_hapmap() instead.",
             FutureWarning,
         )
-        rate_map = read_hapmap(filename)
-        rate = np.append(rate_map.rate, 0)
-        return cls(rate_map.position, rate, map_start=rate_map.map_start)
+        hapmap = np.loadtxt(filename, skiprows=1, usecols=(1, 2))
+        try:
+            filename.seek(0)  # If a file object, we need to read from start again
+        except AttributeError:
+            pass
+        rownames = np.genfromtxt(filename, skip_header=True, usecols=0, dtype="str")
```

Comment:
It should be possible to avoid reading the file twice. How about something like the following?
```python
rownames, position, rate = np.loadtxt(
    filename,
    skiprows=1,
    dtype=[("chrom", object), ("position", int), ("rate", float)],
    usecols=(0,1,2),
    unpack=True
)
```

Reply:
Ah, I didn't realise you could give a dtype to np.loadtxt. Much better. Thanks.